x-pack/plugins/files/public/components/upload_file/util/simple_state_subject.ts

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { merge } from 'lodash';
+import { BehaviorSubject } from 'rxjs';
+
+export class SimpleStateSubject<S extends object = {}> extends BehaviorSubject<S> {
+  constructor(initialState: S) {
+    super(initialState);
+  }
+
+  public getSnapshot() {
+    return this.getValue();
+  }
+
+  public setState(nextState: Partial<S>): void {
+    this.next(merge({}, this.getSnapshot(), nextState));
```

Comment:
could this work without lodash?
```suggestion
    this.next({ ...this.getSnapshot(), ...nextState });
```

Reply:
That could works too! The only difference with that and `merge` is it produces a shallow copy. For the purposes of keeping this "generic" I thought a recursive merge/update would work better.